tests/test_operations.py

```diff
@@ -68,6 +77,103 @@ def test_stage_commit_push_changes__push_failure__raises_pygitops_error(
             )
 
 
+def _delete_existing_file(local_repo: Repo, filename: str) -> None:
+    test_second_file_path = Path(local_repo.working_dir) / filename
+    test_second_file_path.unlink()
+
+
+def _modify_existing_file(local_repo: Repo, filename: str, content: str) -> None:
+    test_file_path = Path(local_repo.working_dir) / filename
+    test_file_path.write_text(content)
+
+
+def _add_new_file(local_repo: Repo, filename: str) -> None:
+    test_other_file_path = Path(local_repo.working_dir) / filename
+    test_other_file_path.touch()
+
+
+@pytest.mark.parametrize(
+    (
+        "some_source_modifier",
+        "some_source_modifier_inputs",
+        "some_other_source_modifier",
+        "some_other_source_modifier_inputs",
+        "some_result",
+        "some_other_result",
+    ),
+    [
+        (
+            _modify_existing_file,
+            [SOME_CONTENT_FILENAME, SOME_CONTENT],
+            _modify_existing_file,
+            [SOME_CONTENT_FILENAME, SOME_CONTENT],
+            SOME_MODIFY_FILE_DIFF,
+            SOME_MODIFY_FILE_DIFF,
+        ),
+        (
+            _delete_existing_file,
+            [SOME_CONTENT_FILENAME],
+            _delete_existing_file,
+            [SOME_CONTENT_FILENAME],
+            SOME_DELETE_FILE_DIFF,
+            SOME_DELETE_FILE_DIFF,
+        ),
+        (
+            _add_new_file,
+            [SOME_OTHER_FILENAME],
+            _add_new_file,
+            [SOME_OTHER_FILENAME],
+            SOME_NEW_FILE_DIFF,
+            SOME_NEW_FILE_DIFF,
+        ),
+        (
+            _modify_existing_file,
+            [SOME_CONTENT_FILENAME, SOME_CONTENT],
+            _add_new_file,
+            [SOME_OTHER_FILENAME],
+            SOME_MODIFY_FILE_DIFF,
+            SOME_NEW_FILE_DIFF,
+        ),
+        (
+            _delete_existing_file,
+            [SOME_CONTENT_FILENAME],
+            _modify_existing_file,
+            [SOME_CONTENT_FILENAME, SOME_CONTENT],
+            SOME_DELETE_FILE_DIFF,
+            SOME_MODIFY_FILE_DIFF,
+        ),
+        (
+            _add_new_file,
+            [SOME_OTHER_FILENAME],
+            _delete_existing_file,
+            [SOME_CONTENT_FILENAME],
+            SOME_NEW_FILE_DIFF,
+            SOME_DELETE_FILE_DIFF,
+        ),
+    ],
```

Comment:
this is awesome!

* I love that you are having the source_modifers be callables
* I love that we get to test a variety of these scenarios!

fantastic job designing these tests and making them this reusable 😄